### PR TITLE
[GTIN] Add GTIN migration JOB

### PR DESCRIPTION
**Pull request by puntope**
### Changes proposed in this Pull Request:

Closes part of #2616 as part of pcTzPl-2p2-p2.

This PR creates the logic for the Job responsible for the migration GTIN in batches for all the products.

### Screenshots:

Screenshot 2024-10-17 at 18 29 40

### Detailed test instructions:

1. Set a simple product and add a GTIN in Google for WooCommerce tab
2. Set a simple product  and add a different GTIN in Google for WooCommerce tab and in the Inventory tab
3. Set a Variable product and set GTIN on one of the variations in Google for WooCommerce  only and in both (inventory and Google for WooCommerce) in other.
4. Go to Connection Test page and click on "Start GTIN Migration"
5. See the job  `gla/jobs/migrate_gtin/create_batch` scheduled in WooCommerce - Status -  Scheduled actions
6. Run the job and see the `gla/jobs/migrate_gtin/process_items` created with the product ids.
7. Run it
8. Go back to the products created in steps 1,2,3 and verify that:
9. For the product in step 1- the GTIN was loaded  in the inventory GTIN
10. For the product in step 2 - The GTIN is still the GTIN you set initially and is not override  by the GTIN in google for WooCommerce GTIN
11. Check the same in the variations (GTIN set in core should not be overrided in the migration).

### Additional details:

In follow-up PRs I will add:

-  the banners for starting the migration
-  tests
-  WP CLI tool

### Changelog entry

> Add - Add GTIN Migration Job

### CODE REVIEW

File: src/Jobs/MigrateGTIN.php [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		return parent::can_schedule( $args ) && $this->is_gtin_available_in_core();
+	}
+
+	/**
+	 * Process batch items.
+	 *
+	 * @param int[] $items A single batch of WooCommerce product IDs from the get_batch() method.
+	 */
+	protected function process_items( array $items ) {
+		// todo: prevent execution if migration was completed?
```
Comment (tomalec) [inline src/Jobs/MigrateGTIN.php:77]: :question:  Is this something you are planning to add in the follow-up PRs, or was it missed?

**Review by tomalec — COMMENTED** [review]
I run the tests locally and for simple products it goes smooth.

:question: 
I noticed a different behavior for simple products when the GTIN is already set to the same value or when it's set to a different one. Is that expected? if there are differences, it keeps the core GTIN but adds the G4W as SKU, while for the case where they're the same, it does not set SKU.
image




Plus, for variable ones, it didn't migrate the GTINs.
The process job was executed.
image
image

File: src/Jobs/MigrateGTIN.php (hunk 1 above)
Comment (puntope) [reply src/Jobs/MigrateGTIN.php:77]: Yes. This is a todo and will be done in a further PR when setting the Banner for the merchant to run the job

**Comment by puntope** [comment]
> I noticed a different behavior for simple products when the GTIN is already set to the same value or when it's set to a different one. Is that expected? if there are differences, it keeps the core GTIN but adds the G4W as SKU, while for the case where they're the same, it does not set SKU.

Hey @tomalec that's odd I was not able to reproduce this. Can you verify again or help me to reproduce this issue? 

The expected behaviour is:

- If GTIN is set in the inventory tab then migration does nothing.
- If GTIN is not set in the inventory tab then migration loads the GTIN set in Google for WooCommerce tab (if it is set)

**Comment by puntope** [comment]
> Plus, for variable ones, it didn't migrate the GTINs. The process job was executed.


The variables are required to process the Variations. I just added a tweak [here](https://github.com/woocommerce/google-listings-and-ads/pull/2645/commits/d39b358d27a404f4a21cc4e57c9da56cc5261f64) because I notice I was processing only available variations (those with price set) Now I do get all the variations.

Notice the GTIN in the variable inventory is useless because we don't support GTIN in Variables. So you need to check each of the variations (they have their own GTIN fields in core part and G4W tab) .

**Comment by puntope** [comment]
Hey @tomalec I addressed all your comments and questions. Can you follow up with another review?

**Review by tomalec — APPROVED** [review]
Thanks for changes and clarifications, now it works well for me :)

**Pull request merged**